Add `all_triangles` generator yielding all unique triangles in a graph

georakom opened the pull request:
Implemented an `all_triangles` function that efficiently yields all unique triangles in an undirected graph.
The algorithm follows a "Forward" principle by enforcing a node ordering to avoid redundant comparisons and double-counting. For each node u, it examines neighbors v > u and yields triangles formed by the intersection of u's and v's neighbors that are greater than v. 
The function optionally supports a nodes parameter, restricting enumeration to triangles where all three nodes are in the specified set.

This is my first pull request here and I would gladly appreciate any comments and improvements!

rossbar commented on networkx/algorithms/cluster.py:
```diff
@@ -247,6 +247,72 @@ def wt(u, v):
+    if nodes is None:
+        # Pre-computation based on fixed ID ordering
+        neighbor_sets = {
+            node: {nbr for nbr in G.neighbors(node) if nbr > node} for node in G
```
This optimization is invalid in networkx - it is never guaranteed that nodes are ordered this way. In fact, it's not even guaranteed that nodes are comparable!

georakom commented:
Thank you for the feedback @rossbar ! I’ve gone ahead and implemented the generator version as all_triangles, following the naming pattern you mentioned.

rossbar reviewed (commented):
Make sure to remove the extraneous changes to the other files that have nothing to do with this PR, which would be required before this gets any in-depth review.

Also - you will likely garner more traction from the community if you can provide stronger motivation for the addition. For example, what problem were you trying to solve where this functionality was necessary? Similarly, why not just use one of the other proposed methods - do we really need a new function for this?

amcandio commented:
This implement is storing all triangles, which means memory consumption is cubical on the amount of nodes in the worst case. I know in NetworkX you cannot depend on ordering the nodes but you can do a linear memory solution.

You assign an incremental id to each node as you see them for the first time. Then you only yield the ones with form (u, v, w) where u < v < w.

georakom commented:
You are right, the implementation did need an improvement. I was thinking of some sort of "visited" ordering, but your idea is simpler to execute, and so I implemented it. The function now assigns an ID to the nodes needed and then computes intersections while avoiding redundant checks. Thank you for the suggestion, @amcandio!

amcandio commented on networkx/algorithms/cluster.py:
```diff
@@ -247,6 +247,80 @@ def wt(u, v):
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    if nodes is None:
+        node_to_id = {node: i for i, node in enumerate(G.nodes())}
```
You can just iterate G

```suggestion
        node_to_id = {node: i for i, node in enumerate(G)}
```

amcandio commented on networkx/algorithms/cluster.py:
```diff
@@ -247,6 +247,80 @@ def wt(u, v):
+    if nodes is None:
+        node_to_id = {node: i for i, node in enumerate(G.nodes())}
+
+        neighbor_sets = {
+            node: {nbr for nbr in G[node] if node_to_id[nbr] > node_to_id[node]}
+            for node in G.nodes()
+        }
+        for u in G.nodes():
+            for v in neighbor_sets[u]:
+                common = neighbor_sets[u] & neighbor_sets[v]
+                for w in common:
+                    yield u, v, w
```
This is consuming n^2 extra memory. You don't need to cache the nodes, for you could do:

```suggestion
        for u in G:
            for v in G[u]:
                if node_to_id[v] <= node_to_id[u]:
                    continue
                for w in G[u] & G[v]:
                    if node_to_id[w] <= node_to_id[v]:
                        continue
                    yield (u, v, w)
```

georakom replied:
Yeah you are right, but for a large graph will they have similar performance?

amcandio replied:
Yes, it could even be faster due to less memory allocations

georakom replied:
Thank you for the suggestion, I went on and applied it!

amcandio commented on networkx/algorithms/cluster.py:
```diff
@@ -254,6 +254,70 @@ def wt(u, v):
+        all_needed = relevant_nodes | {
+            nbr for node in relevant_nodes for nbr in G.neighbors(node)
+        }
+
```
We can dedupe this by only setting an id for relevant nodes. Then you access the id by doing `node_to_id.get(u, -1)`. This way you can use the same routine for both

georakom replied:
Thanks again @amcandio ! You are right, the `all_needed` was a too defensive way to avoid key errors, but this fixes that issue. I went ahead and changed this as well!

amcandio commented on networkx/algorithms/cluster.py:
```diff
@@ -254,6 +254,65 @@ def wt(u, v):
+
+        node_to_id = {node: i for i, node in enumerate(relevant_nodes)}
+
+        for u in relevant_nodes:
+            for v in G[u]:
+                if node_to_id.get(v, -1) <= node_to_id[u]:
+                    continue
+                for w in set(G[u]) & set(G[v]):
+                    if node_to_id.get(w, -1) <= node_to_id[v]:
+                        continue
+                    if u in nodes_set or v in nodes_set or w in nodes_set:
+                        yield u, v, w
```
By both I mean, whether you pass "nodes" or not

georakom replied:
You mean something like this?
```
for u in iterator:
        for v in G[u]:
            if node_to_id.get(v, -1) <= node_to_id[u]:
                continue
            for w in set(G[u]) & set(G[v]):
                if node_to_id.get(w, -1) <= node_to_id[v]:
                    continue
                if nodes_set is None or u in nodes_set or v in nodes_set or w in nodes_set:
                    yield u, v, w
```
I thought of this as well, but I just thought that it might cause a slight overhead if nodes is None, because we do redundant checkings.

dschult replied:
I think what @amcandio meant was you don't need to check that the nodes are in nodes_set anymore because you have bypassed them if they aren't in nodes_set.

Also, this might be a place where `G.nbunch_iter(nodes)` is useful. When the input is a collection of nodes, we typically call it an "nbunch" and feed it to `G.nbunch_iter` which handles `None` and ignores nodes not in `G`. 
`nodes_set = set(G.nbunch_iter(nodes))`

You can also avoid lots of set creation using `w not in unbrs` instead of `&`. It's worth seeing if that's better.
And I added the minor optimization of using `G._adj[u]` instead of `G[u]` which avoids the `Adjacency/Atlas` wrappers -- useful when you know you are only reading the adjacency dict.
```python
for u in nodes_set:
        u_id = node_to_id[u]
        unbrs = set(G._adj[u])
        for v in unbrs:
            v_id = node_to_id.get(v, -1)
            if v_id <= u_id:
                continue
            for w in G._adj[v]:
                if w not in unbrs or (node_to_id.get(w, -1) <= v_id):
                    continue
                yield u, v, w
```
The last `if` is no longer needed because we already checked if v and w are in nodes_set.  It means the whole function is shorter too. :)

amcandio replied:
Sorry, was very unclear. I meant that you don't need to duplicate the 3 nested loops. You can make the second case generalize the first one

amcandio reviewed (approved):
LGTM, great work!

georakom commented:
@amcandio You mean we should make the description more in detail? Or sharpen it?

georakom commented:
@amcandio oh excuse me I thought you were referring to the doc string in the function. Yes I will change the description of the PR. Thank you!

georakom commented on networkx/algorithms/cluster.py:
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    nodes_set = dict.fromkeys(G.nbunch_iter(nodes))
```
But we have to iterate in `nodes_set` either way since we have every option to operate in the same for loop. You mean I should put `nodes_set = set(G)` instead of the dict if nodes are is None?

dschult replied:
I think the suggestion is to have `node_set = G`.  Which also reminds us that `node_set` is not a set anymore and should probably be renamed.  How about something like:
```python
if nodes is None:
    nodes = relevant_nodes = G
else:
    nodes = dict.fromkeys(G.nbunch_iter(nodes))
    relevant_nodes = chain(...
    )
node_to_id = {node: i for i, node in enumerate(relevant_nodes)}

for u in nodes:
    ...
```
And we should probably rename the input parameter from `nodes` to `nbunch` to make it clear that we are using the standard networkx handling of a collection of nodes (nbunch is described in the NetworkX glossary).

Peiffap commented on networkx/algorithms/cluster.py:
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
+        An undirected graph. Self-loops and multiple edges are ignored.
+
+    nodes : node, iterable of nodes, or None (default=None)
+        If a node or iterable of nodes is given, only triangles involving those nodes are yielded.
```
Wording nit: it's unclear from this sentence whether only triangles that have all of their nodes in `nodes` are considered, or only triangles that have at least one of their nodes in `nodes` (or only triangles that have all of the nodes in `nodes`—although that would be an unlikely interpretation...).
```suggestion
        If a node or iterable of nodes is given, only triangles with all their nodes in `nodes` are yielded.
```

georakom replied:
I wasn't very specific with it, I will describe it more concisely now, thank you! However your suggestion doesn't completely apply to what the `nbunch` is actually for. If an iterable of nodes is given, the triangles that are actually yielded are the ones that contain **atleast one** of the nodes in `nbunch`.

Peiffap commented on networkx/algorithms/cluster.py:
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected graph. Self-loops and multiple edges are ignored.
```
Minor nit: I believe this doesn't matter (self-loops and multiedges shouldn't affect the result), and it might just end up confusing users ("will this function work for multigraphs?").
```suggestion
        An undirected graph.
```

Peiffap commented on networkx/algorithms/tests/test_cluster.py:
```diff
@@ -2,6 +2,8 @@
 
 import networkx as nx
 
+from ..cluster import all_triangles
+
```
I think this won't be necessary anymore once the new function is properly exposed in the top-level namespace. You'll be able to just call `nx.all_triangles` then.

Peiffap commented on networkx/algorithms/tests/test_cluster.py:
```diff
@@ -77,6 +79,71 @@ def test_k5(self):
+    assert triangles == [(0, 1, 2)]
+
+
+def test_all_triangles_no_duplicates():
+    G = nx.complete_graph(4)  # 4 nodes, 4-choose-3 = 4 triangles
+    triangles = list(all_triangles(G))
+    assert len(triangles) == 4
```
This is redundant if `test_all_triangles_complete_graph_exact` is kept.

Peiffap commented on networkx/algorithms/tests/test_cluster.py:
```diff
@@ -77,6 +79,71 @@ def test_k5(self):
+
+    assert triangles == expected
+
+
```
Can we add a test where the graph has no triangles too ? And can we add a test where the graph has triangles but none of them in `nodes`?

Some more tests we should consider adding (but that I don't feel are strictly necessary):

- Test that this raises `NetworkXNotImplemented` for directed graphs
- Parametrize the existing tests for `nx.Graph` and `nx.MultiGraph`.

georakom replied:
Yes, I will add these right now!

Peiffap reviewed (commented):
This is nice! Some comments below, but feel free to push back if you disagree :).

Can this be used to speed up `nx.simple_cycles` for the undirected, `length_bound = 3` case? It would be nice to benchmark that... It might also be interesting to add it to the See Also section for that function in that case. This should probably all be done in a separate PR, though, if we decide it's worthwhile.

georakom commented on networkx/algorithms/cluster.py:
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
```
I completely agree. If I am not mistaken, the `node_to_id` ordering is based on how is the graph inserted, right? If I build manually a graph and I do `G.add_nodes_from([2, 0, 5])`, I think the triangle (0,2,5) will be printed like (2,0,5). I am not 100% sure so let me know!

Peiffap replied:
That's actually a separate point I believe—there are two levels at which ordering can matter:

1. This note is about the order in which the tuples are returned, e.g. `all_triangles(G)` could yield `(0, 1, 2)` then `(0, 1, 3)` etc., but yielding them in a different order would still be correct—making the examples and tests use lists means this order is somehow "checked", when I don't believe it should be.
2. The order of nodes inside a tuple is also relevant, and another comment I made addresses that question. A tuple `(u, v, w)` can also be given as `(v, w, u)`, etc. Since technically for an undirected triangle, all 6 node orderings are valid (unlike for a cycle of length 4, where the order matters), this could also be a set, but I think by analogy with `nx.simple_cycles`, an argument can be made that it's perhaps more consistent to return a tuple. I'd be curious to hear other opinions on this.

georakom replied:
The tests and the example have a specific ordering on how the graph is built, and yes I agree that this can give the impression that the triangles are always yielded in a specific order that the human eye sees, which might not be the case. So I think I do need to remove the mention of `u < v < w` from the docstring and make the tests order-independent (using sets or frozensets).

Peiffap replied:
Maybe returning a tuple `(u, v, w)` but running the tests with set (something like `assert {set(t) for t in nx.all_triangles(G)} == {{u1, v1, w1}, {u2, v2, w2}, ...}` makes the most sense then!

georakom commented:
Thank you for all the suggestions @Peiffap ! I went ahead and applied every single one of them.
- Changed the comments and put notes for implementation description and complexity
- Applied set-based comparisons in the tests
- Added the tests you recommended!
I am not very experienced so I added every single suggestion you gave manually just in case it doesn't undo other changes, I hope that's fine :) !

georakom commented on networkx/algorithms/cluster.py:
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
+    Yields
+    -------
+    tuple
+         A tuple of three nodes forming a triangle ``(u, v, w)``.
```
I need to be more careful with such details, thank you!

georakom commented on networkx/algorithms/cluster.py:
```diff
@@ -254,6 +254,63 @@ def wt(u, v):
+    Notes
+    -----
+    This algorithm ensures each triangle is yielded once using an internal node ordering.
+    It runs in O(m * d) time in the worst case, where m is the number of edges and d is the maximum degree.
```
A set intersection of two sets of size at most d costs O(d) and not O(d**2). I am not doing an inner nested scan over both neighbor lists since python’s set intersection does the efficient hashed lookup approach.

georakom commented:
Thank you @Peiffap. I have now added it the `all_triangles` function to `__all__` and also fixed the issues in the comments!

rossbar reviewed (approved):
I took the liberty of pushing up a few minor changes related to formatting etc. but this LGTM! Thanks @georakom and @amcandio + @Peiffap for the detailed review!
